### PR TITLE
Change ra_snapshot:read API to mirror ra_snapshot:accept

### PR DESCRIPTION
**Pull request by hairyhum**
Split ra_snapshot:read/2 to begin_read/2 and read_chunk/2
The goal of this change is to give a clearer API to snapshot
implementers, so they can think of read and accept stages with
similar functions.

### CODE REVIEW

**Comment by hairyhum** [comment]
With ChunkSize being an argument to read_chunk, we can change the chunk size in flight. I'm not sure how valuable it is, but it's possible now.

**Comment by hairyhum** [comment]
I'm not sure why do we need Crc to be a part of the protocol and restrict it's type. `complete_accept` and `validate` functions should validate the data integrity of the snapshot and for `ra_log_snapshot` the Crc is a part of on-disk data. Do we need to transfer it separately? If we do, is it a part of the snapshot metadata?

**Comment by kjnilsson** [comment]
For multi file transfers a single Crc doesn't make sense so we can probably remove this and include a crc as part of the chunk term

File: src/ra_server_proc.erl [hunk 1]
```diff
@@ -1186,31 +1186,35 @@ receive_snapshot_timeout() ->
...
+    Result = read_chunks_and_send_rpc(RPC, To, ReadState, 1, ChunkSize, SnapState),
     ok = gen_statem:cast(Me, {To, Result}).
+
+read_chunks_and_send_rpc(RPC0, To, ReadState0, Num, ChunkSize, SnapState) ->
+    Conf = case Num of
+               1 -> RPC0#install_snapshot_rpc.last_config;
+               _ -> undefined
+           end,
+    {ok, Data, ContState} = ra_snapshot:read_chunk(ReadState0, ChunkSize, SnapState),
+    ChunkFlag = case ContState of
```
Comment (kjnilsson) [inline src/ra_server_proc.erl]: formatting

File: src/ra_snapshot.erl [hunk 2]
```diff
@@ -310,16 +317,25 @@ delete(Dir, {Idx, Term}) ->
...
     Location = make_snapshot_dir(Dir, Idx, Term),
-    Mod:read(ChunkSizeBytes, Location).
+    Mod:begin_read(ChunkSizeBytes, Location).
+
+
+-spec read_chunk(ReadState, ChunkSizeBytes :: non_neg_integer(), State :: state()) ->
+    {ok, Data :: term(), {next, ReadState} | last}  |
+    {error, term()} when ReadState :: term().
+read_chunk(ReadState, ChunkSizeBytes, #?MODULE{module = Mod,
+                                directory = Dir,
```
Comment (kjnilsson) [inline src/ra_snapshot.erl]: align fields

File: src/ra_snapshot.erl [hunk 3]
```diff
@@ -310,16 +317,25 @@ delete(Dir, {Idx, Term}) ->
...
-
--spec read(ChunkSizeBytes :: non_neg_integer(), State :: state()) ->
-    {ok, Crc :: non_neg_integer(), Meta :: meta(), ChunkState,
-     [ChunkThunk :: fun((ChunkState) -> {binary(), ChunkState})]} |
-    {error, term()} when ChunkState :: term().
-read(ChunkSizeBytes, #?MODULE{module = Mod,
+-spec begin_read(ChunkSizeBytes :: non_neg_integer(), State :: state()) ->
+    {ok, Crc :: non_neg_integer(), Meta :: meta(), ReadState} |
+    {error, term()} when ReadState :: term().
+begin_read(ChunkSizeBytes, #?MODULE{module = Mod,
```
Comment (kjnilsson) [inline src/ra_snapshot.erl]: align fields below

**Pull request merged**